src/integrity-checker.js

```diff
@@ -63,51 +63,42 @@ export default class InstallationIntegrityChecker {
 
   config: Config;
 
-  async _getModuleLocation(usedRegistries?: Set<RegistryNames>): Promise<string> {
-    // build up possible folders
-    let registries = registryNames;
-    if (usedRegistries && usedRegistries.size > 0) {
-      registries = usedRegistries;
-    }
-    const possibleFolders = [];
-    if (this.config.modulesFolder) {
-      possibleFolders.push(this.config.modulesFolder);
-    }
-
-    // ensure we only write to a registry folder that was used
-    for (const name of registries) {
-      const loc = path.join(this.config.cwd, this.config.registries[name].folder);
-      possibleFolders.push(loc);
-    }
+  /**
+   * Get the common ancestor of every node_modules - it might be a node_modules directory itself, but isn't required to.
+   */
 
-    // if we already have an integrity hash in one of these folders then use it's location otherwise use the
-    // first folder
-    let loc;
-    for (const possibleLoc of possibleFolders) {
-      if (await fs.exists(path.join(possibleLoc, constants.INTEGRITY_FILENAME))) {
-        loc = possibleLoc;
-        break;
-      }
+  _getModulesRootFolder(): string {
+    if (this.config.modulesFolder) {
+      return this.config.modulesFolder;
+    } else if (this.config.workspaceRootFolder) {
+      return this.config.workspaceRootFolder;
+    } else {
+      return path.join(this.config.lockfileFolder, 'node_modules');
```

Comment:
I'd like to avoid using `this.config.registry`, since its semantic isn't clear (cf #4109). However I can put `node_modules` in the `constants.js` file, if you prefer.